pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,564 @@
+
+	return nil
+}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (b *storageosUnmounter) TearDownAt(dir string) error {
+	if err := util.UnmountPath(dir, b.mounter); err != nil {
+		glog.V(4).Infof("Unmounted StorageOS volume %s failed with: %v", b.pvName, err)
+	}
+	if err := b.manager.UnmountVolume(b); err != nil {
+		glog.V(4).Infof("Mount reference for volume %s could not be removed from StorageOS: %v", b.pvName, err)
+	}
+	return nil
+}
+
+type storageosDeleter struct {
+	*storageos
+}
+
+var _ volume.Deleter = &storageosDeleter{}
+
+func (d *storageosDeleter) GetPath() string {
+	return getPath(d.podUID, d.namespace, d.volName, d.pvName, d.plugin.host)
+}
+
+func (d *storageosDeleter) Delete() error {
+	return d.manager.DeleteVolume(d)
+}
+
+type storageosProvisioner struct {
+	*storageos
+	options volume.VolumeOptions
+}
+
+var _ volume.Provisioner = &storageosProvisioner{}
+
+func (v *storageosProvisioner) Provision() (*v1.PersistentVolume, error) {
+
+	// Set parameters from StorageClass
+	if len(v.options.Parameters) > 0 {
+		if description, ok := v.options.Parameters["description"]; ok {
+			v.description = description
+		}
+		if pool, ok := v.options.Parameters["pool"]; ok {
+			v.pool = pool
+		}
+		if fsType, ok := v.options.Parameters["fstype"]; ok {
+			v.fsType = fsType
+		}
+	}
+
+	// Set from PVC
+	v.volName = v.options.PVName
+	v.namespace = v.options.PVC.Namespace
+	v.labels = make(map[string]string)
+	for k, val := range v.options.PVC.Labels {
+		v.labels[k] = val
+	}
+	capacity := v.options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	v.size = int(volume.RoundUpSize(capacity.Value(), 1024*1024*1024))
+
+	vol, err := v.manager.CreateVolume(v)
+	if err != nil {
+		return nil, err
+	}
+	if vol.FSType == "" {
+		vol.FSType = defaultFSType
+	}
+
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   vol.Namespace + "~" + vol.Name,
+			Labels: map[string]string{},
+			Annotations: map[string]string{
+				"kubernetes.io/createdby": "storageos-dynamic-provisioner",
+			},
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeReclaimPolicy: v.options.PersistentVolumeReclaimPolicy,
+			AccessModes:                   v.options.PVC.Spec.AccessModes,
+			Capacity: v1.ResourceList{
+				v1.ResourceName(v1.ResourceStorage): resource.MustParse(fmt.Sprintf("%dGi", vol.Size)),
+			},
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				StorageOS: &v1.StorageOSVolumeSource{
+					VolumeName: vol.Name,
+					Namespace:  vol.Namespace,
+					Pool:       vol.Pool,
+					FSType:     vol.FSType,
+					ReadOnly:   false,
+				},
+			},
+		},
+	}
+	if len(v.options.PVC.Spec.AccessModes) == 0 {
+		pv.Spec.AccessModes = v.plugin.GetAccessModes()
+	}
+	if len(vol.Labels) != 0 {
+		if pv.Labels == nil {
+			pv.Labels = make(map[string]string)
+		}
+		for k, v := range vol.Labels {
+			pv.Labels[k] = v
```

Comment:
This is setting labels created by StorageOS when the volume is provisioned and applying them to the PV.  So in the example above, the label `storageos.feature.replicas=2` would be set on the PV.  These labels are only informational.